GoogleDataTransport/GDTCORLibrary/GDTCOREventMetricsCounter.m

```diff
@@ -0,0 +1,136 @@
+/*
+ * Copyright 2022 Google LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#import <Foundation/Foundation.h>
+
+#import "GoogleDataTransport/GDTCORLibrary/Private/GDTCOREventMetricsCounter.h"
+
+#import "GoogleDataTransport/GDTCORLibrary/Public/GoogleDataTransport/GDTCOREvent.h"
+
+typedef NSDictionary<NSNumber *, NSNumber *> GDTCORDroppedEventCounter;
+
+@interface GDTCOREventMetricsCounter ()
+
+// TODO(ncooke3): Document.
+@property(nonatomic)
+    NSDictionary<NSString *, GDTCORDroppedEventCounter *> *droppedEventCounterByMappingID;
+
+@end
+
+@implementation GDTCOREventMetricsCounter
+
++ (instancetype)counterWithEvents:(NSArray<GDTCOREvent *> *)events
+                 droppedForReason:(GDTCOREventDropReason)reason {
+  NSMutableDictionary<NSString *, GDTCORDroppedEventCounter *> *eventCounterByMappingID =
+      [NSMutableDictionary dictionary];
+
+  for (GDTCOREvent *event in events) {
+    // TODO(ncooke3): Should I use an autorelease pool?
+    // Dropped events with a `nil` or empty mapping ID (log source) are not recorded.
+    if (event.mappingID.length == 0) {
+      continue;
+    }
+
+    // Get the dropped event counter for the event's mapping ID (log source).
+    // If the dropped event counter for this event's mapping ID is `nil`,
+    // an empty mutable counter is returned.
+    NSMutableDictionary<NSNumber *, NSNumber *> *eventCounter =
+        [NSMutableDictionary dictionaryWithDictionary:eventCounterByMappingID[event.mappingID]];
+
+    // Increment the dropped event counter for the given reason.
+    NSInteger currentEventCountForReason = [eventCounter[@(reason)] integerValue];
+    NSInteger updatedEventCountForReason = currentEventCountForReason + 1;
+
+    eventCounter[@(reason)] = @(updatedEventCountForReason);
```

Comment:
Just curious so for each event there are multiple reasons that can cause it drop and there are multiple types of events as well.

Reply:
Correct! For example, a batch of X events may be dropped for reason Y.